test/e2e/upgrades/apps/deployments.go

```diff
@@ -98,89 +86,82 @@ func (t *DeploymentUpgradeTest) Setup(f *framework.Framework) {
 	})
 	framework.ExpectNoError(err)
 
-	// Use observedGeneration to determine if the controller noticed the pod template update.
-	framework.Logf("Wait deployment %q to be observed by the deployment controller", deploymentName)
-	framework.ExpectNoError(framework.WaitForObservedDeployment(c, ns, deploymentName, deployment.Generation))
-
-	// Wait for it to be updated to revision 2
-	By(fmt.Sprintf("Waiting deployment %q to be updated to revision 2", deploymentName))
-	framework.ExpectNoError(framework.WaitForDeploymentRevisionAndImage(c, ns, deploymentName, "2", nginxImage))
-
 	By(fmt.Sprintf("Waiting deployment %q to complete", deploymentName))
-	framework.ExpectNoError(framework.WaitForDeploymentStatus(c, deployment))
+	framework.ExpectNoError(framework.WaitForDeploymentCompletes(c, deployment))
 
-	rs, err = deploymentutil.GetNewReplicaSet(deployment, c.ExtensionsV1beta1())
+	By(fmt.Sprintf("Getting replicaset of deployment %q after rollout", deploymentName))
+	rsList, err = rsClient.List(metav1.ListOptions{LabelSelector: rsSelector.String()})
 	framework.ExpectNoError(err)
-	if rs == nil {
-		framework.ExpectNoError(fmt.Errorf("expected a new replica set for deployment %q", deployment.Name))
+	rss = rsList.Items
+	if len(rss) != 2 {
+		framework.ExpectNoError(fmt.Errorf("expected 2 replicaset, got %d", len(rss)))
 	}
 
-	if rs.UID == t.oldRS.UID {
-		framework.ExpectNoError(fmt.Errorf("expected a new replica set different from the previous one"))
+	By(fmt.Sprintf("Checking replicaset of deployment %q that is created before rollout survives the rollout", deploymentName))
+	// While checking whether old RS survives rollout, we also set t.newRSUID accordingly.
+	if rss[0].UID == oldRSUID {
+		t.newRSUID = rss[1].UID
+	} else if rss[1].UID == oldRSUID {
+		t.newRSUID = rss[0].UID
+	} else {
+		framework.ExpectNoError(fmt.Errorf("old replicaset with UID %q does not survive rollout", oldRSUID))
 	}
 
-	// Store new replica set - should be the same after the upgrade.
-	t.newRS = rs
-	deployment, err = c.Extensions().Deployments(ns).Get(deployment.Name, metav1.GetOptions{})
+	By(fmt.Sprintf("Verifying revision of the deployment %q is 2", deploymentName))
+	framework.ExpectNoError(framework.WaitForDeploymentRevision(c, deployment, "2"))
+
+	deployment, err = deploymentClient.Get(deploymentName, metav1.GetOptions{})
```

Comment:
You don't need to get deployment again, the first one you got is enough, since all you need is its name, UID, and namespace. Maybe we should only store those information in `t`. You don't actually need `t.updatedD` either.

Reply:
As we utilize the same namespace for upgrade test, it can be obtained from `framework.Namespace.Name`. The only variable requires is the old deployment UID.